Faster optimized frozen dictionary creation (6/6)

adamsitnik opened the pull request:
- don't use custom ToArray for small frozen collections, up to 50% gain for creation time for collections with <= 4 items

|                            Type |             Method |        Job | Count |        Mean | Ratio | Allocated | Alloc Ratio |
|-------------------------------- |------------------- |----------- |------ |------------:|------:|----------:|------------:|
|      Perf_Frozen&lt;ReferenceType&gt; | ToFrozenDictionary |         PR |     4 |    98.70 ns |  0.54 |     152 B |        0.66 |
|      Perf_Frozen&lt;ReferenceType&gt; | ToFrozenDictionary |       main |     4 |   182.50 ns |  1.00 |     232 B |        1.00 |
|                                 |                    |            |       |             |       |           |             |
|      Perf_Frozen&lt;ReferenceType&gt; |        ToFrozenSet |         PR |     4 |   165.18 ns |  0.83 |     384 B |        0.91 |
|      Perf_Frozen&lt;ReferenceType&gt; |        ToFrozenSet |       main |     4 |   198.46 ns |  1.00 |     424 B |        1.00 |

- use the int trick for uint, short, ushort, sbyte and byte: we know that all hash codes are unique and we can avoid some work later. 10-15% CPU time gain and 15-20% allocation reduction for FrozenDictionary and FrozenHashSet where TKey is uint, short, ushort, byte, sbyte

|             Method |  Job | Count |         Mean | Ratio | Allocated | Alloc Ratio |
|------------------- |----- |------ |-------------:|------:|----------:|------------:|
| ToFrozenDictionary |   PR |    64 |  2,808.38 ns |  0.91 |    4400 B |        0.78 |
| ToFrozenDictionary | main |    64 |  3,072.64 ns |  1.00 |    5656 B |        1.00 |
|                    |      |       |              |       |           |             |
|        ToFrozenSet |   PR |    64 |  2,964.31 ns |  0.87 |    5408 B |        0.81 |
|        ToFrozenSet | main |    64 |  3,420.82 ns |  1.00 |    6664 B |        1.00 |
|                    |      |       |              |       |           |             |
| ToFrozenDictionary |   PR |   512 | 25,362.06 ns |  0.87 |   45216 B |        0.84 |
| ToFrozenDictionary | main |   512 | 28,944.59 ns |  1.00 |   53672 B |        1.00 |
|                    |      |       |              |       |           |             |
|        ToFrozenSet |   PR |   512 | 27,882.97 ns |  0.84 |   51632 B |        0.86 |
|        ToFrozenSet | main |   512 | 33,151.36 ns |  1.00 |   60088 B |        1.00 |

- move Length Buckets code to a dedicated helper type to reduce code duplication and decrease code size

@stephentoub this is my last PR for frozen collections perf for now

I apologize if the CI gets red, for some reason I am getting a build error and I temporarily can't run unit tests locally:

```log
CSC : error CS8785: Generator 'JSExportGenerator' failed to generate source. It will not contribute to the output a
nd compilation errors may occur as a result. Exception was of type 'TypeLoadException' with message 'Could not load
 type 'Microsoft.Interop.IGeneratorDiagnostics' from assembly 'Microsoft.Interop.SourceGeneration, Version=42.42.42
.42, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.' [D:\projects\runtime\src\libraries\System.Collections.Immu
table\src\System.Collections.Immutable.csproj::TargetFramework=net7.0]
CSC : error CS8785: Generator 'JSImportGenerator' failed to generate source. It will not contribute to the output a
nd compilation errors may occur as a result. Exception was of type 'TypeLoadException' with message 'Could not load
 type 'Microsoft.Interop.IGeneratorDiagnostics' from assembly 'Microsoft.Interop.SourceGeneration, Version=42.42.42
.42, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.' [D:\projects\runtime\src\libraries\System.Collections.Immu
table\src\System.Collections.Immutable.csproj::TargetFramework=net7.0]
```

contributes to #87964 and closes #87964 it as it's the last PR in the series

adamsitnik committed:
don't use custom ToArray for small frozen collections, up to 50% gain for creation time for collections with <= 4 items

adamsitnik committed:
for these types GetHashCode returns their value casted to int, so when we receive a Dictionary/HashSet where there are key

we know that all hash codes are unique and we can avoid some work later

10-15% CPU time gain and 15-20% allocation reduction for FrozenDictionary and FrozenHashSet where TKey is uint, short, ushort, byte, sbyte

adamsitnik committed:
move Length Buckets code to a dedicated helper type to reduce code duplication and decrease code size

jkoritzinsky commented:
@adamsitnik what commit on main is your branch based on? I believe I fixed the build break you're seeing, so I want to make sure it didn't break in a different way.

stephentoub commented on src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/SmallFrozenDictionary.cs:
```diff
@@ -24,8 +25,8 @@ internal SmallFrozenDictionary(Dictionary<TKey, TValue> source) : base(source.Co
 
-            _keys = source.Keys.ToArray(source.Count);
-            _values = source.Values.ToArray(source.Count);
+            _keys = source.Keys.ToArray();
+            _values = source.Values.ToArray();
```
Isn't this heavily based on the fact that MaxItemsInSmallFrozenCollection is currently 4, which happens to be exactly the default initial array allocation size in ToArray?  If MaxItemsInSmallFrozenCollection gets changed to 8, for example, will this change still be a win for inputs of length 8?

adamsitnik replied:
The custom `ToArray` implementation is slower, because it uses `CopyTo` only for collections that implement `IList<T>`, while both `KeyCollection` and `ValueCollection` don't. They do implement `ICollection<T>` which is used by LINQ.

https://github.com/dotnet/runtime/blob/c06d77a74f64d6c5b0280db56fb0c6a43c0aa907/src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableExtensions.Minimal.cs#L158-L164

https://github.com/dotnet/runtime/blob/c06d77a74f64d6c5b0280db56fb0c6a43c0aa907/src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableExtensions.Minimal.cs#L106

https://github.com/dotnet/runtime/blob/c06d77a74f64d6c5b0280db56fb0c6a43c0aa907/src/libraries/System.Collections/ref/System.Collections.cs#L504

https://github.com/dotnet/runtime/blob/c06d77a74f64d6c5b0280db56fb0c6a43c0aa907/src/libraries/System.Collections/ref/System.Collections.cs#L532

stephentoub replied:
> because it uses CopyTo only for collections that implement IList<T>

We should revisit this.
https://github.com/dotnet/runtime/blob/e5ddc6c9b7e481d36d6350960868188d7b275ef7/src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableExtensions.Minimal.cs#L93-L96
That logic, while well-meaning, isn't worth making the implementations slower.  If code wants to get at the underlying array of an ImmutableArray, there are easier ways, including a now fully supported public API.

adamsitnik replied:
I agree, but I don't believe it's in the scope of this PR, so I've created a dedicated issue for it: https://github.com/dotnet/runtime/issues/88181

stephentoub commented on src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/Constants.cs:
```diff
@@ -78,5 +78,13 @@ internal static class Constants
+
+        // for these types GetHashCode returns their value casted to int, so when we receive a Dictionary/HashSet where there are key
+        // we know that all hash codes are unique and we can avoid some work later
+        internal static bool KeysAreHashCodes<T>()
+            => typeof(T) == typeof(int) || typeof(T) == typeof(uint)
+            || typeof(T) == typeof(short) || typeof(T) == typeof(ushort)
+            || typeof(T) == typeof(byte) || typeof(T) == typeof(sbyte)
+            || ((typeof(T) == typeof(IntPtr) || typeof(T) == typeof(UIntPtr)) && IntPtr.Size == 4);
```
Do we have tests that use all of these types as keys?

adamsitnik replied:
Great catch, we did not have the coverage. I added the coverage and fixed the existing tests (some of them were just lucky to work due to what random numbers were generated).

stephentoub commented on src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/Constants.cs:
```diff
@@ -78,5 +78,13 @@ internal static class Constants
+
+        // for these types GetHashCode returns their value casted to int, so when we receive a Dictionary/HashSet where there are key
+        // we know that all hash codes are unique and we can avoid some work later
+        internal static bool KeysAreHashCodes<T>()
+            => typeof(T) == typeof(int) || typeof(T) == typeof(uint)
+            || typeof(T) == typeof(short) || typeof(T) == typeof(ushort)
+            || typeof(T) == typeof(byte) || typeof(T) == typeof(sbyte)
+            || ((typeof(T) == typeof(IntPtr) || typeof(T) == typeof(UIntPtr)) && IntPtr.Size == 4);
```
nit: nint and nuint?

gfoidl replied:
These are already handled by `{U]IntPtr`. See [sharplab](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEBMBGAWAChis8BOACgDEB7WgHgEkYEAFBZAPkoEpeA3KQo16DAKot2nHvyFEyVOoxiwEswcKViYqNRvnCAbAAJg9ADYnlDACqziAb2InXJrAHYTCAJ4AHAFNaADNKW14TAF5I738g0KkOZAiMDFjAkLCI6PT4yklWJM0iAF8gA=) for demo.

stephentoub replied:
No, I meant weet should write "nint" and "nuint" instead of "IntPtr" abd "UIntPtr". It was a style comment.

adamsitnik committed:
Merge remote-tracking branch 'upstream/main' into minorFrozenPerfFixes

adamsitnik committed:
address code review feedback:

* use nint and nuint rather than IntPtr and UIntPtr
* add tests for Frozen Dictionaries with key being uint, short, ushort, byte, sbyte

adamsitnik commented on src/libraries/Common/tests/System/Collections/ICollection.Generic.Tests.cs:
```diff
@@ -406,8 +406,10 @@ public void ICollection_Generic_Contains_ValidValueOnCollectionContainingThatVal
         {
             ICollection<T> collection = GenericICollectionFactory(count);
-            if (DefaultValueAllowed)
+            if (DefaultValueAllowed && default(T) is null) // it's true only for reference types
```
0 is a perfect key for a dictionary of int

stephentoub replied:
> it's true only for reference types

And for `Nullable<T>`

adamsitnik commented on src/libraries/Common/tests/System/Collections/IDictionary.Generic.Tests.cs:
```diff
@@ -289,7 +289,7 @@ public void IDictionary_Generic_ItemGet_MissingNonDefaultKey_ThrowsKeyNotFoundEx
         public void IDictionary_Generic_ItemGet_MissingDefaultKey_ThrowsKeyNotFoundException(int count)
         {
-            if (DefaultValueAllowed)
+            if (DefaultValueAllowed && !IsReadOnly)
```
this test was working only because existing int-based tests  did not generate 0 as a key:

https://github.com/dotnet/runtime/blob/c06d77a74f64d6c5b0280db56fb0c6a43c0aa907/src/libraries/Common/tests/System/Collections/IDictionary.Generic.Tests.cs#L295-L297

adamsitnik commented on src/libraries/System.Collections.Immutable/tests/Frozen/FrozenDictionaryTests.cs:
```diff
@@ -392,19 +406,68 @@ public void CreateHugeDictionary_Success(int largeCount)
+
+    public class FrozenDictionary_Generic_Tests_int_int : FrozenDictionary_Generic_Tests_base_for_numbers<int>
+    {
+        protected override int Next(Random random) => random.Next(); 
```
the existing test was using `Next` overload that does not generate negative values. I decided to keep it that way

adamsitnik commented:
>  I believe I fixed the build break you're seeing, so I want to make sure it didn't break in a different way.

@jkoritzinsky I've synced my branch with upstream and everything works fine. Thanks for letting me know!

adamsitnik committed:
address code review feedback: improve the comment

stephentoub commented on src/libraries/System.Collections.Immutable/tests/Frozen/FrozenDictionaryTests.cs:
```diff
@@ -392,19 +406,68 @@ public void CreateHugeDictionary_Success(int largeCount)
+        protected override byte Next(Random random) => (byte)random.Next(byte.MinValue, byte.MaxValue);
+    }
+
+    public class FrozenDictionary_Generic_Tests_sbyte_sbyte : FrozenDictionary_Generic_Tests_base_for_numbers<sbyte>
```
You stopped short of creating tests for nint and nuint?

adamsitnik replied:
> You stopped short of creating tests for nint and nuint? Seems like those could be some of the more valuable ones, since the optimization is only relevant when they're 4 bytes.

It was an oversight. I added the tests and fixed the bug it had discovered (`IntPtr` started implementing `IComparable<IntPtr>` in .NET 5)

adamsitnik committed:
address code review feedback: add nint and uint test coverage

and also fix a discovered bug: IntPtr started implementing IComparable<IntPtr> in .NET 5

stephentoub commented on src/libraries/System.Collections.Immutable/tests/Frozen/FrozenDictionaryTests.cs:
```diff
@@ -421,6 +421,17 @@ public abstract class FrozenDictionary_Generic_Tests_base_for_numbers<T> : Froze
+            byte[] bytes = new byte[8];
+            random.NextBytes(bytes);
+            return BitConverter.ToInt64(bytes, 0);
+#endif
```
It's not going to matter, but technically these aren't the same thing.  NextInt64(long.MinValue, long.MaxValue) will never produce long.MaxValue, but the else code can.

adamsitnik replied:
> will never produce long.MaxValue, but the else code can. 

Good point! After thinking about it I decided to remove the NET6_0_OR_GREATER case as it could cause another kind of issues: generate different output for the same input and hence get us to a situation where exactly the same test is passing for one moniker and failing for another.

adamsitnik committed:
address code review feedback: remove NET6_0_OR_GREATER case

adamsitnik commented:
The failure is unrelated (#87477), merging.
